File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/planning/RelationshipJoinOrderingRule.java
package com.foundgine.core.semantic.planning;

import com.foundgine.core.semantic.RelationshipCardinality;
import com.foundgine.core.semantic.query.*;
import java.util.*;

/** Assigns a deterministic physical traversal order to eligible sibling relationship nodes. */
public final class RelationshipJoinOrderingRule implements IPlanRewriteRule {
    @Override public String name() { return "relationship.join.order"; }
    @Override public List<String> preconditions() { return List.of("plan contains two or more sibling relationship traversals", "relationship cardinality is known for each candidate traversal", "no explicit sibling execution order is already fixed"); }
    @Override public List<String> securityObligations() { return List.of("authorization.required", "authorization.runtime", "visibility.relationship", "planning.cache-context-isolation"); }
    @Override public double costImpact() { return 1d; }
    @Override public double benefitEstimate() { return 2.5d; }
    @Override public int priority() { return 30; }
    @Override public boolean canApply(SemanticPlan plan) { Objects.requireNonNull(plan); return containsEligibleSiblingSet(plan.root()); }
    @Override public SemanticPlan apply(SemanticPlan plan) {
        Objects.requireNonNull(plan); if (!canApply(plan)) return plan;
        boolean[] changed = {false}; var root = rewrite(plan.root(), changed);
        return changed[0] ? new SemanticPlan(root, plan.requiredSecurityInvariants(), plan.authorizationBinding()) : plan;
    }
    private static SemanticPlanNode rewrite(SemanticPlanNode node, boolean[] changed) {
        List<SemanticPlanNode> children = new ArrayList<>(node.children());
        var candidates = new ArrayList<Integer>();
        for (int i = 0; i < children.size(); i++) {
            var child = children.get(i);
            if (child.viaRelationship() != null && child.relationshipCardinality() != null && child.traversalOrder() < 0) candidates.add(i);
        }
        if (candidates.size() >= 2) {
            candidates.sort(Comparator.comparingInt((Integer i) -> selectivityClass(children.get(i)))
                    .thenComparingInt(i -> children.get(i).relationshipCardinality() == RelationshipCardinality.ONE ? 0 : 1)
                    .thenComparingLong(i -> children.get(i).viaRelationship().value()).thenComparingInt(Integer::intValue));
            for (int rank = 0; rank < candidates.size(); rank++) {
                int index = candidates.get(rank); var child = children.get(index);
                if (child.traversalOrder() != rank) {
                    children.set(index, copyWithTraversalOrder(child, rank)); changed[0] = true;
                }
            }
        }
        boolean childChanged = false;
        for (int i = 0; i < children.size(); i++) { var before = children.get(i); var after = rewrite(before, changed); if (after != before) { children.set(i, after); childChanged = true; } }
        if (childChanged) changed[0] = true;
        return childChanged || !children.equals(node.children()) ? copyWithChildren(node, children) : node;
    }
    private static int selectivityClass(SemanticPlanNode node) {
        int score = 0;
        var filter = node.queryOptions() == null ? null : node.queryOptions().filter();
        if (filter != null) score -= 3;
        if (filter instanceof SemanticRelationshipFilter || filter instanceof SemanticAggregateFilter) score -= 1;
        if (node.queryOptions() != null && node.queryOptions().limit() != null && node.queryOptions().limit() > 0) score -= 2;
        if (node.relationshipCardinality() == RelationshipCardinality.ONE) score -= 1;
        return score;
    }
    private static boolean containsEligibleSiblingSet(SemanticPlanNode node) {
        long count = node.children().stream().filter(c -> c.viaRelationship() != null && c.relationshipCardinality() != null && c.traversalOrder() < 0).count();
        return count >= 2 || node.children().stream().anyMatch(RelationshipJoinOrderingRule::containsEligibleSiblingSet);
    }
    private static SemanticPlanNode copyWithTraversalOrder(SemanticPlanNode n, int order) { return new SemanticPlanNode(n.id(),n.operation(),n.entityId(),n.fields(),n.viaRelationship(),n.viaConnection(),n.children(),n.queryOptions(),n.authorization(),n.relationshipCardinality(),n.traversalMode(),order,n.aggregateExecutionStrategy()); }
    private static SemanticPlanNode copyWithChildren(SemanticPlanNode n,List<SemanticPlanNode> children) { return new SemanticPlanNode(n.id(),n.operation(),n.entityId(),n.fields(),n.viaRelationship(),n.viaConnection(),children,n.queryOptions(),n.authorization(),n.relationshipCardinality(),n.traversalMode(),n.traversalOrder(),n.aggregateExecutionStrategy()); }
}

File: src/csharp/Foundgine.Core/Semantic/Planning/RelationshipJoinOrderingRule.cs
using Foundgine.Core.Semantic;
using Foundgine.Core.Semantic.Query;

namespace Foundgine.Core.Semantic.Planning;

/// <summary>
/// Assigns a deterministic physical traversal order to sibling relationship
/// nodes. The logical child collection is not reordered, so result shaping and
/// requested field order remain unchanged. Providers may use the order when
/// constructing an execution strategy, subject to their own semantic and
/// security conformance checks.
/// </summary>
public sealed class RelationshipJoinOrderingRule : IPlanRewriteRule
{
    public string Name => "relationship.join.order";

    public IReadOnlyList<string> Preconditions =>
    [
        "plan contains two or more sibling relationship traversals",
        "relationship cardinality is known for each candidate traversal",
        "no explicit sibling execution order is already fixed"
    ];

    public IReadOnlyList<string> SecurityObligations =>
    [
        "authorization.required",
        "authorization.runtime",
        "visibility.relationship",
        "planning.cache-context-isolation"
    ];

    public double CostImpact => 1.0d;
    public double BenefitEstimate => 2.5d;
    public bool IsIdempotent => true;
    public int Priority => 30;

    public bool CanApply(SemanticPlan plan)
    {
        ArgumentNullException.ThrowIfNull(plan);
        return ContainsEligibleSiblingSet(plan.Root);
    }

    public SemanticPlan Apply(SemanticPlan plan)
    {
        ArgumentNullException.ThrowIfNull(plan);
        if (!CanApply(plan))
            return plan;

        var changed = false;
        var root = Rewrite(plan.Root, ref changed);
        return changed ? new SemanticPlan(root, plan.RequiredSecurityInvariants, plan.AuthorizationBinding) : plan;
    }

    private static SemanticPlanNode Rewrite(SemanticPlanNode node, ref bool changed)
    {
        var children = node.Children.ToArray();
        var childChanged = false;

        var candidates = children
            .Select((child, index) => (child, index))
            .Where(x => x.child.ViaRelationship is not null &&
                        x.child.RelationshipCardinality is not null)
            .ToArray();

        if (candidates.Length >= 2 && candidates.All(x => x.child.TraversalOrder < 0))
        {
            var ordered = candidates
                .OrderBy(x => SelectivityClass(x.child))
                .ThenBy(x => x.child.RelationshipCardinality == RelationshipCardinality.One ? 0 : 1)
                .ThenBy(x => x.child.ViaRelationship!.Value.Value)
                .ThenBy(x => x.index)
                .ToArray();

            for (var rank = 0; rank < ordered.Length; rank++)
            {
                var (child, index) = ordered[rank];
                var updated = child with { TraversalOrder = rank };
                if (!ReferenceEquals(updated, child))
                {
                    children[index] = updated;
                    childChanged = true;
                }
            }
        }

        for (var i = 0; i < children.Length; i++)
        {
            var updated = Rewrite(children[i], ref changed);
            if (!ReferenceEquals(updated, children[i]))
            {
                children[i] = updated;
                childChanged = true;
            }
        }

        if (childChanged)
        {
            changed = true;
            return node with { Children = children };
        }

        return node;
    }

    private static int SelectivityClass(SemanticPlanNode node)
    {
        var score = 0;
        if (node.QueryOptions?.Filter is not null)
            score -= 3;
        if (node.QueryOptions?.Filter is SemanticRelationshipFilter or SemanticAggregateFilter)
            score -= 1;
        if (node.QueryOptions?.Limit is > 0)
            score -= 2;
        if (node.RelationshipCardinality == RelationshipCardinality.One)
            score -= 1;
        return score;
    }

    private static bool ContainsEligibleSiblingSet(SemanticPlanNode node)
    {
        var count = node.Children.Count(child =>
            child.ViaRelationship is not null &&
            child.RelationshipCardinality is not null &&
            child.TraversalOrder < 0);

        return count >= 2 || node.Children.Any(ContainsEligibleSiblingSet);
    }
}
